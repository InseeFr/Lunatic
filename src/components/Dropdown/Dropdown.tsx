import type { LunaticComponentProps } from '../type';
import { getComponentErrors } from '../shared/ComponentErrors/ComponentErrors';
import { slottableComponent } from '../shared/HOC/slottableComponent';
import { SimpleOptionRenderer } from './renderer/SimpleOptionRenderer';
import { SimpleLabelRenderer } from './renderer/SimpleLabelRenderer';
import { Combobox } from '../shared/Combobox/Combobox';
import classNames from 'classnames';
import type { LunaticError } from '../../use-lunatic/type';

export function Dropdown({
	handleChange,
	response,
	errors,
	...props
}: LunaticComponentProps<'Dropdown'>) {
	return (
		<CustomDropdown
			{...props}
			onChange={(v) => handleChange(response, v)}
			errors={getComponentErrors(errors, props.id)}
		/>
	);
}

type CustomProps = Omit<
	LunaticComponentProps<'Dropdown'>,
	'response' | 'handleChange' | 'errors'
> & {
	onChange: (v: string | null) => void;
	errors?: LunaticError[];
};

export const CustomDropdown = slottableComponent<CustomProps>(
	'Dropdown',
	(props) => {
		const {
			id,
			disabled,
			options,
			className,
			value,
			description,
			label,
			errors,
			onChange,
			readOnly,
		} = props;

		return (
			<Combobox
				id={id}
				className={classNames(className, 'lunatic-dropdown')}
				disabled={disabled}
				readOnly={readOnly}
				options={options}
				editable={false}
				onSelect={onChange}
				optionRenderer={SimpleOptionRenderer}
				labelRenderer={SimpleLabelRenderer}
				value={value}
				label={label}
				errors={errors}
				description={description}
			/>
		);
	}
);
