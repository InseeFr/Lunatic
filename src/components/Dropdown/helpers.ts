import removeAccents from 'remove-accents';

/**
 * Filters an array of options based on a search prefix.
 **/
export function filterOptions<T extends { label?: unknown; search?: string }>(
	options: T[],
	prefix?: string
): T[] {
	if (!prefix || prefix.length === 0) {
		return options;
	}
	return options
		.map((option) => ({
			option,
			score: matchScore(option.search || labelToString(option.label), prefix),
		}))
		.filter((option) => option.score >= 0.6)
		.sort((a, b) => b.score - a.score)
		.map((o) => o.option);
}

/**
 * Extract string content from the variable received
 */
export function labelToString(label: unknown): string {
	if (label && typeof label === 'object' && 'props' in label) {
		const { props } = label;
		const { expression } = props as Record<string, string>;
		return `${expression}`;
	}

	return `${label}`;
}

/**
 * Finds letters in a label that match a given prefix.
 **/
export function lettersMatching(label: string, prefix: string = ''): string[] {
	const pref = prefix.split('');
	let pos = -1;
	const clean = preparePrefix(label);
	return clean.split('').reduce((a: string[], c: string) => {
		if (pos < pref.length && pref[pos + 1] === c) {
			pos++;
			return [...a, c];
		}
		return a;
	}, []);
}

/**
 * Calculates a score for matching a label against a prefix.
 **/
function matchScore(label: string, prefix: string): number {
	const pref = prefix.split('');
	let pos = -1;
	const clean = preparePrefix(label);
	return (
		clean.split('').reduce((a, c, idx) => {
			if (pos < pref.length && pref[pos + 1] === c) {
				pos++;
				return a + 1;
			}
			return a / (idx * 0.0001 + 1);
		}, 0) / prefix.length
	);
}

/**
 * Sanitize a prefix for search
 */
export function preparePrefix(prefix: unknown): string {
	if (typeof prefix === 'string') {
		return removeAccents(prefix.toLowerCase()).replace(/[- ']/g, '');
	}
	return '';
}
