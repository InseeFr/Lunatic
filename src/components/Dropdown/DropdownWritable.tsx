import { type ReactNode, useCallback, useState } from 'react';
import type { ComboboxOptionType } from '../shared/Combobox/ComboboxType';
import type { LunaticError } from '../../use-lunatic/type';
import { WritableOptionRenderer } from './renderer/WritableOptionRenderer';
import { WritableLabelRenderer } from './renderer/WritableLabelRenderer';
import { Combobox } from '../shared/Combobox/Combobox';
import { filterOptions } from './helpers';

type Props = {
	id?: string;
	disabled?: boolean;
	options: ComboboxOptionType[];
	onSelect: (v: string | null) => void;
	className?: string;
	value: string | null;
	label?: ReactNode;
	errors?: LunaticError[];
	description?: ReactNode;
	readOnly?: boolean;
};

export function DropdownWritable({
	id,
	disabled,
	options,
	onSelect,
	className,
	value,
	label,
	errors,
	description,
	readOnly,
}: Props) {
	const [filtered, setFiltered] = useState(options);

	const onChange = useCallback(
		function (search: string | null) {
			if (search) {
				setFiltered(filterOptions(options, search));
			} else {
				setFiltered(options);
			}
			onSelect(null);
		},
		[options, onSelect]
	);

	return (
		<Combobox
			id={id}
			className={className}
			disabled={disabled}
			readOnly={readOnly}
			options={filtered}
			onSelect={onSelect}
			onChange={onChange}
			optionRenderer={WritableOptionRenderer}
			labelRenderer={WritableLabelRenderer}
			editable={true}
			value={value}
			label={label}
			errors={errors}
			description={description}
		/>
	);
}
