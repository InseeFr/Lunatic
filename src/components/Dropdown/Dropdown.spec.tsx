import { render } from '@testing-library/react';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import { Dropdown } from './Dropdown';

describe('Dropdown', () => {
	const mockOnSelect = vi.fn();

	const baseProps = {
		value: 'toto',
		id: 'dropdown',
		'aria-labelledby': 'dropdown',
		options: [
			{
				value: 'toto',
				description: 'toto',
				label: 'toto',
			},
		],
		response: { name: 'demo' },
		handleChanges: mockOnSelect,
	};

	beforeEach(() => {
		mockOnSelect.mockClear();
	});

	it('renders without crashing', () => {
		const { container } = render(<Dropdown {...baseProps} />);
		expect(container).toMatchSnapshot();
	});

	it('should handle readOnly', () => {
		const { container } = render(<Dropdown {...baseProps} readOnly />);
		expect(container).toMatchSnapshot();

		const selection = container.querySelector('.lunatic-combo-box-content');
		(selection as HTMLElement).focus();
		expect(selection).toHaveFocus();
		const span = selection?.querySelector('span');
		expect(span).toHaveTextContent('toto');
	});
});
