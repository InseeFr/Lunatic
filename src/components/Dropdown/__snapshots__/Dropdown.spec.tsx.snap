// Vitest Snapshot v1

exports[`Dropdown > renders without crashing 1`] = `
<div>
  <div
    class="lunatic-dropdown lunatic-combo-box-container lunatic-suggester-default-style lunatic-suggester-default-style default-style"
    id="lunatic-combo-box-container-dropdown"
  >
    <div
      class="lunatic-combo-box"
    >
      <div
        class="lunatic-combo-box-content"
        tabindex="0"
      >
        <div
          aria-autocomplete="list"
          aria-controls="todo"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="label-dropdown"
          aria-owns="dropdown"
          class="lunatic-combo-box-selection"
          id="dropdown"
          role="combobox"
        >
          <span
            class="selection"
          >
            toto
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`Dropdown > should handle readOnly 1`] = `
<div>
  <div
    class="lunatic-dropdown lunatic-combo-box-container lunatic-suggester-default-style lunatic-suggester-default-style default-style"
    id="lunatic-combo-box-container-dropdown"
  >
    <div
      class="lunatic-combo-box"
    >
      <div
        class="lunatic-combo-box-content"
        tabindex="0"
      >
        <div
          aria-autocomplete="list"
          aria-controls="todo"
          aria-expanded="false"
          aria-haspopup="listbox"
          aria-labelledby="label-dropdown"
          aria-owns="dropdown"
          class="lunatic-combo-box-selection"
          id="dropdown"
          role="combobox"
        >
          <span
            class="selection"
          >
            toto
          </span>
        </div>
      </div>
    </div>
  </div>
</div>
`;
